File: IT/2/Programmazione funzionale/Esercizi/C++/esercizio24.cpp
/*
file esercizio24.cpp
compilare con: g++ -std=c++11 esercizio24.cpp
*/

#include <iostream>
#include <functional>

using std::cout;
using std::endl;
using std::bind;
using namespace std::placeholders;

int main() {
	auto f = [] (const int a, const int b) { return a * b; };

	cout << "3 x 3 = " << f(3, 3) << endl;

	auto g = bind(f, 1, _1);

	cout << "id(3) = " << g(3) << endl;

	auto h = [] (const int a, const int b) { return (double) a / b; };

	cout << "3 / 2 = " << h(3, 2) << endl;

	// calcola l'elemento inverso

	auto i = bind(h, 1, _1);

	cout << "Inverso moltiplicativo di 3 = " << i(3) << endl;
}

File: IT/2/Programmazione funzionale/Esercizi/Python/esercizio24.py
# file esercizio24.py

from __future__ import print_function, division
from functools import partial

f = lambda a, b: a * b

print("3 x 3 = {}".format(f(3, 3)))

g = partial(f, 1)

print("id(3) = {}".format(g(3)))

h = lambda a, b : a / b

print("3 / 2 = {}".format(h(3, 2)))

# calcola l'elemento inverso

i = partial(h, 1)

print("Inverso moltiplicativo di 3 = {:.6f}".format(i(3)))

